python/packages/kagent-adk/src/kagent_adk/_token.py

```diff
@@ -0,0 +1,80 @@
+import asyncio
+from contextlib import asynccontextmanager
+from typing import Optional
+
+KAGENT_TOKEN_PATH = "/var/run/secrets/tokens/kagent-token"
+
+
+class KAgentTokenService:
+    """Reads a k8s token from a file, and reloads it
+    periodically.
+    """
+
+    def __init__(self, app_name: str):
+        self.token = None
+        self.update_lock = asyncio.Lock()
+        self.update_task = None
+        self.app_name = app_name
+
+    def lifespan(self):
+        """Returns an async context manager to start the token update loop"""
+
+        @asynccontextmanager
+        async def _lifespan(app: any):
+            await self._update_token_loop()
+            yield
+            self._drain()
+
+        return _lifespan
+
+    def event_hooks(self):
+        """Returns a dictionary of event hooks for the application
+        to use when creating the httpx.AsyncClient.
+        """
+        return {"request": [self._add_bearer_token]}
+
+    async def _update_token_loop(self) -> str:
+        self.token = await self._read_kagent_token()
+        # keep it updated - launch a background task to refresh it periodically
+        self.update_task = asyncio.create_task(self._refresh_token())
+
+    def _drain(self):
+        if self.update_task:
+            self.update_task.cancel()
+
+    async def _get_token(self) -> str:
+        async with self.update_lock:
+            return self.token
+
+    async def _read_kagent_token(self) -> Optional[str]:
+        token = await asyncio.to_thread(read_token)
+        if not token:
+            return None
+        return token
+
+    async def _refresh_token(self):
+        while True:
+            await asyncio.sleep(60)  # Wait for 60 seconds before refreshing
+            token = await self._read_kagent_token()
+            if token is not None and token != self.token:
+                async with self.update_lock:
+                    # python is not multi threaded, so this might not even need a lock.
+                    self.token = token
+
+    async def _add_bearer_token(self, request):
+        # Your function to generate headers dynamically
+        token = await self._get_token()
+        headers = {"X-Agent-Name": self.app_name}
+        if token:
+            headers["Authorization"] = f"Bearer {token}"
+        request.headers.update(headers)
+
+
+def read_token() -> str:
+    try:
+        with open(KAGENT_TOKEN_PATH, "r", encoding="utf-8") as f:
+            token = f.read()
+            return token.strip()
+    except OSError:
+        # TODO: figure out how to log - print(f"Error reading token from {KAGENT_TOKEN_PATH}: {e}")
```

Comment:
This method should change the signature or return "". Also, to add log you can just add a logger at the top of the file: `logger - logging.get(__name__)`